### DISCUSSION TITLE
obj_grad function

### DISCUSSION THREAD

**Discussion by Peter230655** [csu-hmc/opty#177] [Q&A]
In the example of the pedalling with variable h you use this gradient function:

    def obj_grad(free):
        T = free[laenge * num_nodes:]
        grad = np.zeros_like(free)
        grad[laenge * num_nodes:] = 2.0*free[-1]*free[laenge*num_nodes:]
        grad[-1] = np.sum(T**2)
        return grad

The second to last line: grad[-1] = np.sum(T**2)  I do not understand.
What is the 'mathematical' background?

**Thanks for any explanation!**

**Comment by moorepants** [comment]
In #164 this is the objective and its gradient:

```python
def obj(free):
    """Return h (always the last element in the free variables)."""
    return free[-1]


def gradient(free):
    """Return the gradient of the objective."""
    grad = np.zeros_like(free)
    grad[-1] = 1.0
    return grad
```

**Reply by moorepants** [reply]
The objective here is simply minimize h. h is the last value in free (always if variable duration problem). The gradient of h wrt to each variable in free is zero except for the last entry which is dhdh = 1.

**Reply by Peter230655** [reply]
> The objective here is simply minimize h. h is the last value in free (always if variable duration problem). The gradient of h wrt to each variable in free is zero except for the last entry which is dhdh = 1.

The total 'amount of torque needed does not matter here, correct?
I must think about this some more, I think I mixed up mnimizing h and minimizimg 'total torque'.

**Reply by moorepants** [reply]
In the pedaling problem I am minimizing the "race time". I want the cyclist to pass the finish line in the minimal amount of time and they can use their muscles maximally. I'm not minimizing the inputs in this case, I'm simply minimizing time to travel a specific distance.

**Reply by Peter230655** [reply]
> In the pedaling problem I am minimizing the "race time". I want the cyclist to pass the finish line in the minimal amount of time and they can use their muscles maximally. I'm not minimizing the inputs in this case, I'm simply minimizing time to travel a specific distance.

I simply have not given the objective much thought. This discussion helped!

**Comment by Peter230655** [comment]
Clear, thanks!
The example I copied has what you show, but multiplied by sum(T**2).
I will remove, and see if this helps.

**Reply by moorepants** [reply]
You can't just remove things. The gradient has to calculate the gradient of the objective wrt all values in free.

**Comment by Peter230655** [comment]
Now I simply bounded h: 1.e-5 < h < 1, and it finds a solution quickly!
(I am also optimizing h only, and bound the force like -250 < force < 250.)

**Comment by Peter230655** [comment]
my simulation is a double pendulum attached to a chart, which moves left / right on a horizontal line. The force pushes the chart.
In the solution, the force seems to jump from minimum to maximun and back.
I vaguely remember such a "min / max Theorem" from control theory 40+ years ago - or is my memory failing me?

**Reply by moorepants** [reply]
Bang bang control
